zetaclient/chains/ton/observer/inbound.go

```diff
@@ -0,0 +1,260 @@
+	if err != nil {
+		return err
+	}
+
+	var (
+		chainID         = ob.Chain().ChainId
+		initialInterval = ticker.SecondsFromUint64(ob.ChainParams().InboundTicker)
+		sampledLogger   = ob.Logger().Inbound.Sample(&zerolog.BasicSampler{N: 10})
+	)
+
+	ob.Logger().Inbound.Info().Msgf("WatchInbound started for chain %d", chainID)
+
+	task := func(ctx context.Context, t *ticker.Ticker) error {
+		if !app.IsInboundObservationEnabled() {
+			sampledLogger.Info().Msgf("WatchInbound: inbound observation is disabled for chain %d", chainID)
+			return nil
+		}
+
+		if err := ob.observeInbound(ctx); err != nil {
+			ob.Logger().Inbound.Err(err).Msg("WatchInbound: observeInbound error")
+		}
+
+		newInterval := ticker.SecondsFromUint64(ob.ChainParams().InboundTicker)
+		t.SetInterval(newInterval)
+
+		return nil
+	}
+
+	return ticker.Run(
+		ctx,
+		initialInterval,
+		task,
+		ticker.WithStopChan(ob.StopChannel()),
+		ticker.WithLogger(ob.Logger().Inbound, "WatchInbound"),
+	)
+}
+
+func (ob *Observer) observeInbound(ctx context.Context) error {
+	if err := ob.ensureLastScannedTX(ctx); err != nil {
+		return errors.Wrap(err, "unable to ensure last scanned tx")
+	}
+
+	lt, hashBits, err := liteapi.TransactionHashFromString(ob.LastTxScanned())
+	if err != nil {
+		return errors.Wrapf(err, "unable to parse last scanned tx %q", ob.LastTxScanned())
+	}
+
+	txs, err := ob.client.GetTransactionsUntil(ctx, ob.gateway.AccountID(), lt, hashBits)
+	if err != nil {
+		return errors.Wrap(err, "unable to get transactions")
+	}
+
+	// Process from oldest to latest (ASC)
+	slices.Reverse(txs)
+
+	switch {
+	case len(txs) == 0:
+		// noop
+		return nil
+	case len(txs) > MaxTransactionsPerTick:
+		ob.Logger().Inbound.Info().
+			Msgf("observeInbound: got %d transactions. Taking first %d", len(txs), MaxTransactionsPerTick)
+
+		txs = txs[:MaxTransactionsPerTick]
+	default:
+		ob.Logger().Inbound.Info().Msgf("observeInbound: got %d transactions", len(txs))
+	}
+
+	for i := range txs {
+		tx := txs[i]
+
+		parsedTX, skip, err := ob.gateway.ParseAndFilter(tx, toncontracts.FilterInbounds)
+		if err != nil {
+			return errors.Wrap(err, "unable to parse and filter tx")
+		}
+
+		if skip {
+			ob.Logger().Inbound.Info().Fields(txLogFields(&tx)).Msg("observeInbound: skipping tx")
+			ob.setLastScannedTX(&tx)
+
+			continue
+		}
+
+		if _, err := ob.voteInbound(ctx, parsedTX); err != nil {
+			ob.Logger().Inbound.
+				Error().Err(err).
+				Fields(txLogFields(&tx)).
+				Msg("observeInbound: unable to vote for tx")
+
+			return errors.Wrapf(err, "unable to vote for inbound tx %s", tx.Hash().Hex())
+		}
+
+		ob.setLastScannedTX(&parsedTX.Transaction)
+	}
+
+	return nil
+}
+
+func (ob *Observer) voteInbound(ctx context.Context, tx *toncontracts.Transaction) (string, error) {
+	// noop
+	if tx.Operation == toncontracts.OpDonate {
+		ob.Logger().Inbound.Info().
+			Uint64("tx.lt", tx.Lt).
+			Str("tx.hash", tx.Hash().Hex()).
+			Msg("Thank you rich folk for your donation!")
```

Comment:
NIT : Consider making this log line constant which can be used for all observers